web/src/lib/components/photos-page/asset-grid.svelte

```diff
@@ -111,8 +113,80 @@
     navigateToNextAsset();
     assetStore.removeAsset(asset.id);
   };
+
+  let lastAssetMouseEvent: AssetResponseDto | null = null;
+
+  $: if (!lastAssetMouseEvent) {
+    assetInteractionStore.clearAssetSelectionCandidates();
+  }
+
+  let shiftKeyIsDown = false;
+
+  const onKeyDown = (e: KeyboardEvent) => {
+    if (e.key === 'Shift') {
+      e.preventDefault();
+      shiftKeyIsDown = true;
+    }
+  };
+
+  const onKeyUp = (e: KeyboardEvent) => {
+    if (e.key === 'Shift') {
+      e.preventDefault();
+      shiftKeyIsDown = false;
+    }
+  };
+
+  $: if (!shiftKeyIsDown) {
+    assetInteractionStore.clearAssetSelectionCandidates();
+  }
+
+  $: if (shiftKeyIsDown && lastAssetMouseEvent) {
+    selectAssetCandidates(lastAssetMouseEvent);
+  }
+
+  const getLastSelectedAsset = () => {
```

Comment:
Can you help explaning this method?

Reply:
So, I wanted the selection to start from the most recently selected asset. There is no "proper" way to get the most recently added item from a set, and so we have to essentially convert it to an array and get the last element. This function is equivalent to:

```ts
const s = new Set();
[...s].pop();
```

Though, this implementation is much quicker and uses much less memory.